# 01 Daten zusammenfügen
In Pandas können wir DataFrames auf verschiedene Arten zusammenfügen:
- **concat:** Fügt DataFrames aneinander – entweder Zeile für Zeile oder Spalte für Spalte.
- **merge:** Kombiniert DataFrames basierend auf gemeinsamen Spalten, ähnlich wie ein "inner join“ in einer Datenbank. 

In [1]:
# PANDAS IMPORTIEREN
import pandas as pd

### Daten laden
Laden der beiden Files.

In [4]:
# DATAFRAMES LADEN
df_person = pd.read_csv('06_person_daten.csv')
df_adress = pd.read_csv('06_adress_daten.csv')

df_adress

ID    PLZ       Wohnort
0   1  98765    Petersburg
1   2  65432       München
2   3  67543       Hamburg
3   4  87654   Musterstadt

### Merge
Zusammenfügen der beiden Tabellen.

In [7]:
# DATAFRAMES MERGEN
df_merged = pd.merge(df_person, df_adress, on='ID', how='inner')
df_merged

ID  Vorname     Nachname    PLZ       Wohnort
0   1   Tobias       Lämmle  98765    Petersburg
1   2     Paul     Pronobis  65432       München
2   3     Jane          Doe  67543       Hamburg
3   4      Max   Mustermann  87654   Musterstadt

### Daten anhängen
Neue Daten an bestehende Daten anhängen.

In [8]:
# DATAFRAME LADEN
df_person_neu = pd.read_csv("06_person_neu.csv")
df_person_neu

ID  Vorname  Nachname
0   5    Maria    Helwig

In [10]:
# DATAFRAME ANHÄNGEN
df_person2 = pd.concat([df_person, df_person_neu], ignore_index=True)
df_person2

ID  Vorname     Nachname
0   1   Tobias       Lämmle
1   2     Paul     Pronobis
2   3     Jane          Doe
3   4      Max   Mustermann
4   5    Maria       Helwig

# Aufgabe
Lösche Zeile 4 (4, 87654, Musterstadt) aus df_adress. Versuche jetzt noch einmal df_person und df_adress mit merge zusammenzufügen 🔗, jedoch ohne die Personendaten zu "Max Mustermann" zu verlieren. Frage den Chatbot, wie du das machen kannst.

In [12]:
df_adress2 = df_adress.copy()
df_adress2 = df_adress2.drop(index=3)
df_adress2

ID    PLZ      Wohnort
0   1  98765   Petersburg
1   2  65432      München
2   3  67543      Hamburg

In [19]:
df_merged2 = pd.merge(df_person, df_adress2, on='ID', how='left')
df_merged2

ID  Vorname     Nachname      PLZ      Wohnort
0   1   Tobias       Lämmle  98765.0   Petersburg
1   2     Paul     Pronobis  65432.0      München
2   3     Jane          Doe  67543.0      Hamburg
3   4      Max   Mustermann      NaN          NaN